app/vminsert/netstorage/netstorage.go

```diff
@@ -586,14 +632,59 @@ func initStorageNodes(addrs []string, hashSeed uint64) *storageNodesBucket {
 		wg:        &wg,
 	}
 
-	for idx, sn := range sns {
+	// add broken nodes to the end of the list
+	// this is needed because consistent hash slots will be populated with IDs of available
+	// storage nodes (if there are any) and indexes of consistent hash must be linked to healthy storage nodes
+	snb.sns = append(snb.sns, brokenNodes...)
+
+	for idx, sn := range snb.sns {
 		wg.Add(1)
 		go func(sn *storageNode, idx int) {
 			sn.run(snb, idx)
 			wg.Done()
 		}(sn, idx)
 	}
 
+	// Watch for node become healthy and rebuild snb.
+	for _, sn := range brokenNodes {
+		wg.Add(1)
+		go func(sn *storageNode) {
+			defer wg.Done()
+			for {
+				sn.brLock.Lock()
```

Comment:
this code uses multiple mutex locks (sn.brlock + sn.brCond) + atomic pointer loads.

 It opens wide possibility to data races, race conditions, dead locks. 

 It's better to rewrite it and add tests.

Reply:
It's better to write it with:
1) move logic to the separate function
2) add comments if needed. 
3) add tests